File: library/src/main/java/com/alamkanak/weekview/drawing/DrawingContext.java
package com.alamkanak.weekview.drawing;

import com.alamkanak.weekview.model.WeekViewConfig;
import com.alamkanak.weekview.utils.DateUtils;

import java.util.Calendar;
import java.util.List;

public class DrawingContext {

    public final List<Calendar> dayRange;

    //public final int leftDaysWithGaps;
    //public final float totalDayWidth;
    public final float startPixel;

    private DrawingContext(List<Calendar> dayRange, /*int leftDaysWithGaps,
                           float totalDayWidth,*/ float startPixel) {
        this.dayRange = dayRange;
        //this.leftDaysWithGaps = leftDaysWithGaps;
        //this.totalDayWidth = totalDayWidth;
        this.startPixel = startPixel;
    }

    public static DrawingContext create(WeekViewConfig config) {
        final WeekViewDrawingConfig drawConfig = config.drawingConfig;
        final float totalDayWidth = drawConfig.widthPerDay + config.columnGap;
        final int leftDaysWithGaps = (int) (Math.ceil(drawConfig.currentOrigin.x / totalDayWidth) * -1);
        final float startPixel = drawConfig.currentOrigin.x
                + totalDayWidth * leftDaysWithGaps
                + drawConfig.headerColumnWidth;

        final int start = leftDaysWithGaps + 1;
        final int end = start + config.numberOfVisibleDays + 1;
        final List<Calendar> dayRange = DateUtils.getDateRange(start, end);

        return new DrawingContext(dayRange, /*leftDaysWithGaps, totalDayWidth,*/ startPixel);
    }

}

File: core/src/main/java/com/alamkanak/weekview/DrawingContext.kt
package com.alamkanak.weekview

import java.util.Calendar
import kotlin.math.ceil

internal class DrawingContext(
    private val config: WeekViewConfigWrapper
) {

    private var startPixel = 0f

    val startPixels = mutableListOf<Float>()
    val dateRange = mutableListOf<Calendar>()
    val dateRangeWithStartPixels = mutableListOf<Pair<Calendar, Float>>()

    fun update() {
        val totalDayWidth = config.totalDayWidth
        val originX = config.currentOrigin.x

        val daysFromOrigin = ceil(originX / totalDayWidth).toInt() * (-1)
        startPixel = config.timeColumnWidth + originX + totalDayWidth * daysFromOrigin

        val start = daysFromOrigin + 1
        val end = start + config.numberOfVisibleDays

        // If the user is scrolling, a new view becomes partially visible, so we must add an
        // additional date to the date range
        val isNotScrolling = originX % config.totalDayWidth == 0f
        val modifiedEnd = if (isNotScrolling) end - 1 else end

        dateRange.clear()
        dateRange += getDateRange(start, modifiedEnd)

        updateStartPixels()

        dateRangeWithStartPixels.clear()
        dateRangeWithStartPixels += dateRange.zip(startPixels)
    }

    private fun updateStartPixels() {
        startPixels.clear()
        startPixels += dateRange.indices.map {
            index -> startPixel + index * config.totalDayWidth
        }
    }
}
